### DISCUSSION TITLE
xarray groupby does not drop dimension when squeeze

### DISCUSSION THREAD

**Discussion by ecimoli** [pydata/xarray#6173] [Q&A] [answered]
I have MxNxT xarray with the following dimensions and coordinates:

```
<xarray.DataArray (gridcell: 238, window: 9, traits: 3)>
dask.array<where, shape=(238, 9, 3), dtype=float32, chunksize=(14, 9, 3), chunktype=numpy.ndarray>
Coordinates:
    band         (traits) int32 dask.array<chunksize=(3,), meta=np.ndarray>
    spatial_ref  int32 0
  * traits       (traits) <U7 'trait_1' 'trait_2' 'trait_3'
  * gridcell     (gridcell) MultiIndex
  - latitude     (gridcell) float64 6.054e+06 6.054e+06 ... 6.054e+06 6.054e+06
  - longitude    (gridcell) float64 6.042e+05 6.043e+05 ... 6.043e+05 6.043e+05
  * window       (window) MultiIndex
  - x_window     (window) int64 0 0 0 1 1 1 2 2 2
  - y_window     (window) int64 0 1 2 0 1 2 0 1 2
Attributes:
    scale_factor:  1.0
    add_offset:    0.0
```
I am trying to apply groupby the "gridcell" dimension so I can perform an operation on each of those groups using xr.map, or apply_ufunc function.

`kernel_group = kernels_xr.groupby('gridcell', squeeze=True)`

When I do that, the resulting group matrix is 1xNxT, but I only want groups to be NxT, isn't that what squeeze is supposed to do? I don't see any changes whatsoever. Any other suggestions are very much appreciated as well.

Your help would be and immense relief!

Thanks

**Comment by mathause** [comment]
> I am trying to apply groupby the "traits" dimension so I can perform an operation on each of those groups using xr.map function.

> `kernel_group = kernels_xr.groupby('gridcell', squeeze=True)`

Shouldn't the call be `kernel_group = kernels_xr.groupby('traits', squeeze=True)`?

You are currently grouping over `gridcell` - can you check if your `gridcell` actually form groups with more than one member? If not there is nothing to group by...

```python
import numpy as np
import xarray as xr

air = xr.tutorial.open_dataset("air_temperature")
a = air.stack(gridcell=("lat", "lon"))
len(np.unique(a.gridcell.values)) == len(a.gridcell.values)
```

**Reply by ecimoli** [reply]
Apologizes mathause, my typo, i want to group by gridcell, and is indeed forming the groups that i desired. Figured out that re assign the coordinates to a non-multindex and float64 vector, it works and drops the specified dimension, but i lose the multindex stacked information.
Any idea? Thanks again for the time and support in replying

**Comment by mathause** [comment, marked as answer]
Here is a small example. Kind of works 

```python
import numpy as np
import xarray as xr

data = np.arange(8).reshape(2, 4)
da = xr.DataArray(data, dims=("y", "x"), coords=dict(y=[1, 2], x=[1, 1, 1, 1]))

da = da.stack(gridcell=("x", "y"))

da_g = da.groupby("gridcell").mean()

```

It keeps the multiindex but renames the levels (this could be considered a bug).

```
<xarray.DataArray (gridcell: 2)>
array([1.5, 5.5])
Coordinates:
  * gridcell          (gridcell) MultiIndex
  - gridcell_level_0  (gridcell) int64 1 1
  - gridcell_level_1  (gridcell) int64 1 2
```

This can be fixed like so:

```python
mi = da_g.indexes["gridcell"]
da_g.assign_coords(gridcell=mi.set_names(["x", "y"]))
```

Do you experience something different? If so a small example would be helpful. (There is a refactor of the way indexes are treated underway - this may change how this behaves).

**Reply by ecimoli** [reply]
Thanks mate! That did work out as a solution! So much appreciated.

Xarray is magic, hope it keeps improving!

Emiliano
